core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveEmptyMergeWriterRuleSet.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.TableFinishNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import io.trino.sql.tree.GenericLiteral;
+import io.trino.sql.tree.Row;
+
+import java.util.Set;
+
+import static io.trino.sql.planner.plan.Patterns.emptyValues;
+import static io.trino.sql.planner.plan.Patterns.exchange;
+import static io.trino.sql.planner.plan.Patterns.mergeWriter;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableFinish;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * If the predicate for a mergeWriter is optimized to false, the target table scan
+ * of the mergeWriter will be replaced with an empty values node. This type of
+ * plan cannot be executed and is meaningless anyway, so we replace the
+ * entire thing with a values node.
+ * <p>
+ * Transforms
+ * <pre>
+ *  - TableFinish
+ *    - Exchange (optional)
+ *      - MergeWriter
+ *        - Exchange
+ *          - empty Values
+ * </pre>
+ * into
+ * <pre>
+ *  - Values (0)
+ * </pre>
+ */
+public final class RemoveEmptyMergeWriterRuleSet
```

Comment:
Can you add a test for this? Maybe `TestRemoveEmptyDeleteRuleSet` is a good example

Reply:
This turned out to be somewhat challenging.  I spent an hour on it and didn't get the test working.  I'll add the test in a second pass.